### DISCUSSION TITLE
How to run action for the struct which is a parent of the other

### DISCUSSION THREAD

**Discussion by qawbecrdtey** [taocpp/PEGTL#194] [Q&A] [answered]
The title may seem strange, so let me give an example.
```
#include <iostream>
#include <tao/pegtl.hpp>
struct a : tao::pegtl::one<'a'> {};
struct b : a {};

struct A : tao::pegtl::one<'A'> {};
struct B : tao::pegtl::seq<A, tao::pegtl::one<';'>> {};

template<typename Rule>
struct action : tao::pegtl::nothing<Rule> {};

template<>
struct action<a> {
    static void apply0() { std::cout << "a\n" << std::endl; }
};

template<>
struct action<A> {
    static void apply0() { std::cout << "A\n" << std::endl; }
};

int main() {
    std::string s1 = "a";
    std::string s2 = "A;";
    tao::pegtl::memory_input in1(s1, "");
    tao::pegtl::memory_input in2(s2, "");
    tao::pegtl::parse<b, action>(in1);
    tao::pegtl::parse<B, action>(in2);
}
```
The program prints `A` only.
If I'm right, the optimization somewhat tries to use `b` instead of `a`, and discards it so that we can't even use that for an action.
There might be a case like, if we define structs like `list_a`, `list_b`, up to `list_z`, and some of them are combined with other rules while others itself are parents, just like 'a'. One might choose to use a macro, and just like me, find out the action doesn't work.
Then how can I use `a` for an action in this case?

**Comment by d-frey** [comment, marked as answer]
The short answer is: You can't. It's a limitation of the language, as the specialization does not work for direct base classes.

As a work-around you may simply avoid deriving directly and instead use a `seq` as a wrapper. If you use

    struct b : tao::pegtl::seq< a > {};

it should work.

**Comment by qawbecrdtey** [comment]
Thanks for the fast answer!
